File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode() # 虛擬頭節點
        tail = dummy

        p1, p2 = list1, list2

        while p1 and p2:
            if p1.val < p2.val:
                tail.next = p1
                p1 = p1.next
            else:
                tail.next = p2
                p2 = p2.next
            tail = tail.next
        # 將剩餘節點接上
        tail.next = p1 if p1 else p2

        return dummy.next

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummy(0);// 建立虛擬頭節點
        ListNode* tail = &dummy; // 尾指標指向 dummy

        while(list1 && list2){
            if (list1->val < list2->val){
                tail->next = list1;
                list1 = list1->next;
        } else{
            tail->next = list2;
            list2 = list2->next;
        }
        tail = tail->next;
    }

    // 接上剩餘節點
    tail->next = list1 ? list1: list2;
    return dummy.next; // 回傳虛擬頭節點的下一個節點，也就是合併後串列的頭
    }
};
